src/main/kotlin/org/elm/ide/intentions/AddQualifierIntention.kt

```diff
@@ -0,0 +1,126 @@
+import com.intellij.psi.PsiElement
+import org.elm.lang.core.psi.*
+import org.elm.lang.core.psi.elements.*
+import org.elm.lang.core.psi.elements.Flavor.*
+import org.elm.lang.core.resolve.ElmReferenceElement
+import org.elm.lang.core.resolve.scope.ModuleScope
+import org.elm.lang.core.resolve.scope.VisibleNames
+import org.elm.lang.core.types.moduleName
+import org.elm.openapiext.isUnitTestMode
+import org.elm.openapiext.runWriteCommandAction
+import org.elm.openapiext.toPsiFile
+import org.jetbrains.annotations.TestOnly
+
+class AddQualifierIntention : ElmAtCaretIntentionActionBase<AddQualifierIntention.Context>() {
+
+    data class Context(
+            val candidates: List<String>,
+            val referenceName: String,
+            val qid: ElmQID
+    )
+
+    override fun getText() = "Qualify name"
+    override fun getFamilyName() = text
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        val file = editor.toPsiFile(project) as? ElmFile ?: error("no file: should not happen")
+        if (element.ancestors.any { it is ElmModuleDeclaration || it is ElmPortAnnotation || it is ElmImportClause }) {
+            return null
+        }
+
+        val qid = element.parentOfType<ElmQID>() ?: return null
+        if (qid.isQualified) return null
+
+        val typeRef = qid.parentOfType<ElmTypeRef>()
+        if (typeRef != null) {
+            return makeContext(file, qid, typeRef, ModuleScope.getReferencableTypes(file))
+        }
+
+        val valueRef = qid.parentOfType<ElmValueExpr>() ?: return null
+
+        return when (valueRef.flavor) {
+            QualifiedValue, QualifiedConstructor -> null
+            BareValue -> {
+                makeContext(file, qid, valueRef, ModuleScope.getReferencableValues(file))
+            }
+            BareConstructor -> {
+                makeContext(file, qid, valueRef, ModuleScope.getReferencableConstructors(file))
+            }
+        }
+    }
+
+    private fun makeContext(file: ElmFile, qid: ElmQID, ref: ElmReferenceElement, names: VisibleNames): Context? {
+        val name = ref.referenceName
+        val candidates = (names.global + names.imported)
+                .filter { it.name == name }
+                .mapNotNull { ModuleScope.getQualifierForName(file, it.moduleName, name) }
+                .filter { it.isNotEmpty() }
+        if (candidates.isEmpty()) return null
+        return Context(candidates, name, qid)
+    }
+
+    override fun invoke(project: Project, editor: Editor, context: Context) {
+        when (context.candidates.size) {
+            0 -> error("should not happen: must be at least one candidate")
+            1 -> addQualifier(project, context, context.candidates.first())
+            else -> promptToSelectCandidate(project, editor, context)
+        }
+    }
+
+    private fun promptToSelectCandidate(project: Project, editor: Editor, context: Context) {
```

Comment:
this method and `addQualifier` should be swapped so that it reads top-down the same way that you read the `when` in `invoke`